### PR TITLE
Deferred service provider

### LINKED ISSUE #150: Add support for deferred service provider for Laravel >=5.8
In laravel 5.8, `$defer` property of the service provider has been replaced by the `Illuminate\Contracts\Support\DeferrableProvider` interface.

Since the `laravel-settings` package has not implemented that interface for laravel 5.8, its initialization is not deferred.

### PR DESCRIPTION
**Pull request by sausin**
Closes #150

### CODE REVIEW

**Review by bweston92 — CHANGES_REQUESTED** [review]
I don't know if this solution is the best way to go about it, what would you think about having a base class like `class ServiceProvider extends \Illuminate\Support\ServiceProvider` it can keep the `protected $defer = true;` and consumers below Laravel 5.8 can use that but then we have a new file with a class `class DeferrableServiceProvider extends ServiceProvider implements \Illuminate\Contracts\Support\DeferrableProvider` that consumers with Laravel 5.8 and above can utilise?

**Comment by sausin** [comment]
@bweston92 That can definitely be an option. We will either have a new service provider for the interface. But then, for those that are manually importing the service provider in their projects (>=5.5) a different `ServiceProvider` file would be a breaking change. If we have a new file for the older versions, then the package will be break for older laravel versions.

@anlutro it would be great to get your thoughts as well !

File: src/ServiceProvider.php [hunk 1]
```diff
@@ -11,80 +11,134 @@
...
-	 *
-	 * @var boolean
-	 */
-	protected $defer = true;
-
-	/**
-	 * Register IoC bindings.
-	 */
-	public function register()
+if (interface_exists('Illuminate\Contracts\Support\DeferrableProvider') {
```
Comment (bweston92) [inline src/ServiceProvider.php:14]: line looks to be missing a ending brace?

**Comment by bweston92** [comment]
@sausin how are consumers using this without the deferrable interface on the service provider?

If we want to keep it to the same class name, we would need to create a base class and then has the following for the service provider.

```php
if (interface_exists('Illuminate\Contracts\Support\DeferrableProvider'))
  class ServiceProvider extends BaseServiceProvider implements \Illuminate\Contracts\Support\DeferrableProvider {}
else
  class ServiceProvider extends BaseServiceProvider {}
```

**Comment by sausin** [comment]
> @sausin how are consumers using this without the deferrable interface on the service provider?
> 
> If we want to keep it to the same class name, we would need to create a base class and then has the following for the service provider.
> 
> ```
> if (interface_exists('Illuminate\Contracts\Support\DeferrableProvider'))
>   class ServiceProvider extends BaseServiceProvider implements \Illuminate\Contracts\Support\DeferrableProvider {}
> else
>   class ServiceProvider extends BaseServiceProvider {}
> ```

Apologies for the delay in getting back. Without the `DeferableProvider` there is no difference wrt usage, it's only about [performance](https://laravel.com/docs/8.x/providers#deferred-providers).

Did you mean to suggest that we will basically move the interface logic to a `BaseServiceProvider` which will keep the `ServiceProvider` class clean in a way?

I'll was going to update this PR on the weekend. If you're okay to take it on before then please feel free do so. :+1:

**Comment by bweston92** [comment]
Ok so there is no break changes, it would mean the consumers would need to just update to new version and use the `DeferrableServiceProvider`.

So if you leave the service provider as was and do `class DeferrableServiceProvider extends ServiceProvider implements \Illuminate\Contracts\Support\DeferrableProvider` I think that would be the best way, and update the documentation to use this service provider when using Laravel 5.8 or above.

Thanks, have a good weekend.

**Comment by sausin** [comment]
@bweston92 I've updated the PR with hopefully agreeable changes. Edits were made within Github's editor so I hope no typo's got left behind.

**Comment by bweston92** [comment]
Not quite what I was saying, but it should work. I will merge and do a test run when I can. Thanks for your contribution @sausin

**Review by bweston92 — APPROVED** [review]

**Pull request merged**